applications/ColossalEval/colossal_eval/models/vllm.py

```diff
@@ -0,0 +1,534 @@
+            if self.tokenizer.eos_token:
+                self.tokenizer.pad_token = self.tokenizer.eos_token
+            elif hasattr(self.tokenizer, "eod_id"):
+                # Qwen has an eod token "<|endoftext|>".
+                self.tokenizer.pad_token_id = self.tokenizer.eod_id
+
+    def _calculate_loss(self, inputs: List[str], labels: List[str]) -> Tuple[List]:
+        """
+        Calculate loss on target tokens. Adapted from https://github.com/open-compass/opencompass/blob/c2bcd8725e615ec455bf5b7301f8d09962cd64e3/opencompass/models/vllm.py#L110
+
+        Args:
+            input_ids_list: A batch of input string.
+            labels: A batch of labels.
+
+        Returns:
+            A list of loss and a list of label length.
+
+        """
+        batch_size = len(inputs)
+        sampling_kwargs = SamplingParams(logprobs=1)
+        outputs = self.model.generate(inputs, sampling_kwargs)
+        ce_loss = []
+
+        if labels is not None:
+            lens = [len(self.tokenizer.encode(label, add_special_tokens=False)) for label in labels]
+        else:
+            lens = [1] * batch_size
+
+        for i in range(batch_size):
+            logprobs = outputs[i].outputs[0].logprobs
+            token_ids = outputs[i].outputs[0].token_ids
+
+            logprobs_list = [logprobs[i][token_ids[i]] for i in range(len(logprobs))]
+            logprobs_list = [i.logprob for i in logprobs_list]
+            logprobs_list = np.array(logprobs_list)
+
+            if lens is not None:
+                logprobs_list = logprobs_list[: lens[i]]
+
+            loss = -logprobs_list.sum(axis=-1) / lens[i]
+            ce_loss.append(loss)
+
+        batch_loss = np.array(ce_loss)
+
+        return batch_loss, lens
+
+    def _get_truncated_prompts(self, inputs: List[str], max_new_tokens: int) -> List[str]:
+        """
+        Truncate the input sequence to fit model_max_length (we suggest truncate in the middle, since the left and right side may contain crucial instructions)
+        https://github.com/THUDM/LongBench/blob/main/pred.py#L16
+
+        Args:
+            inputs: A batch of input prompts.
+            max_new_tokens: Max new tokens for model to generate.
+
+        Returns:
+            Truncated prompts.
+
+        """
+
+        truncated_inputs = copy.deepcopy(inputs)
+        for i, input in enumerate(inputs):
+            tokenized_prompt = self.tokenizer(input, truncation=False, return_tensors="pt").input_ids[0]
+            if len(tokenized_prompt) > self.model_max_length - max_new_tokens:
+                half = (self.model_max_length - max_new_tokens) // 2
+                prompt = self.tokenizer.decode(
+                    tokenized_prompt[:half], skip_special_tokens=True
+                ) + self.tokenizer.decode(tokenized_prompt[-half:], skip_special_tokens=True)
+                truncated_inputs[i] = prompt
+
+        return truncated_inputs
```

Comment:
As most of the functions can be reused, you should implement your part from `HuggingFaceModel` rather than base model to avoid replicate code.

Please have a look at https://github.com/hpcaitech/ColossalAI/blob/main/applications/ColossalEval/colossal_eval/models/chatglm.py.

Reply:
I think actually most of functions are different from the `HuggingFaceModel` except for the two prompt processing functions `_get_choices_indices` and `_get_truncated_prompts`, even the model & tokenizer loadings of vllm are different. Is it better to inherit from `HuggingFaceModel` for these two functions?

Reply:
Yes, I think so.

As vllm models are also compatible with huggingface so it is better to inherit from `HuggingFaceModel` rather than from base.